File: library/db/src/main/java/org/keynote/godtools/android/db/AttachmentMapper.java
package org.keynote.godtools.android.db;

import android.content.ContentValues;
import android.database.Cursor;
import android.support.annotation.NonNull;

import org.cru.godtools.model.Attachment;
import org.cru.godtools.model.Tool;

import static org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_DOWNLOADED;
import static org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_FILENAME;
import static org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_LOCALFILENAME;
import static org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_SHA256;
import static org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_TOOL;

final class AttachmentMapper extends BaseMapper<Attachment> {
    @Override
    protected void mapField(@NonNull final ContentValues values, @NonNull final String field,
                            @NonNull final Attachment attachment) {
        switch (field) {
            case COLUMN_TOOL:
                values.put(field, attachment.getToolId());
                break;
            case COLUMN_FILENAME:
                values.put(field, attachment.getFileName());
                values.put(COLUMN_LOCALFILENAME, attachment.getLocalFileName());
                break;
            case COLUMN_SHA256:
                values.put(field, attachment.getSha256());
                values.put(COLUMN_LOCALFILENAME, attachment.getLocalFileName());
                break;
            case COLUMN_DOWNLOADED:
                values.put(field, attachment.isDownloaded());
                break;
            default:
                super.mapField(values, field, attachment);
                break;
        }
    }

    @NonNull
    @Override
    protected Attachment newObject(@NonNull final Cursor c) {
        return new Attachment();
    }

    @NonNull
    @Override
    public Attachment toObject(@NonNull final Cursor c) {
        final Attachment attachment = super.toObject(c);

        attachment.setToolId(getLong(c, COLUMN_TOOL, Tool.INVALID_ID));
        attachment.setFileName(getString(c, COLUMN_FILENAME, null));
        attachment.setSha256(getString(c, COLUMN_SHA256, null));
        attachment.setDownloaded(getBool(c, COLUMN_DOWNLOADED, false));

        return attachment;
    }
}

File: library/db/src/main/kotlin/org/keynote/godtools/android/db/AttachmentMapper.kt
package org.keynote.godtools.android.db

import android.content.ContentValues
import android.database.Cursor
import org.ccci.gto.android.common.db.AbstractMapper
import org.ccci.gto.android.common.util.database.getLong
import org.ccci.gto.android.common.util.database.getString
import org.cru.godtools.model.Attachment
import org.cru.godtools.model.Base
import org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_DOWNLOADED
import org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_FILENAME
import org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_LOCALFILENAME
import org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_SHA256
import org.keynote.godtools.android.db.Contract.AttachmentTable.COLUMN_TOOL
import org.keynote.godtools.android.db.Contract.BaseTable.Companion.COLUMN_ID

internal object AttachmentMapper : AbstractMapper<Attachment>() {
    override fun mapField(values: ContentValues, field: String, obj: Attachment) {
        when (field) {
            COLUMN_ID -> values.put(field, obj.id)
            COLUMN_TOOL -> values.put(field, obj.toolId)
            COLUMN_FILENAME -> {
                values.put(field, obj.filename)
                values.put(COLUMN_LOCALFILENAME, obj.localFilename)
            }
            COLUMN_SHA256 -> {
                values.put(field, obj.sha256)
                values.put(COLUMN_LOCALFILENAME, obj.localFilename)
            }
            COLUMN_DOWNLOADED -> values.put(field, obj.isDownloaded)
            else -> super.mapField(values, field, obj)
        }
    }

    override fun newObject(c: Cursor) = Attachment()
    override fun toObject(c: Cursor) = super.toObject(c).apply {
        id = c.getLong(COLUMN_ID, Base.INVALID_ID)
        toolId = c.getLong(COLUMN_TOOL)
        filename = c.getString(COLUMN_FILENAME)
        sha256 = c.getString(COLUMN_SHA256)
        isDownloaded = getBool(c, COLUMN_DOWNLOADED, false)
    }
}
